server/sonar-server/src/main/java/org/sonar/server/qualitygate/QgateProjectFinder.java

```diff
@@ -53,45 +110,35 @@ public boolean hasMoreResults() {
     }
   }
 
-  private final QualityGateDao qualitygateDao;
-  private final ProjectQgateAssociationDao associationDao;
+  private enum ToProjectId implements Function<ProjectQgateAssociationDto, Long> {
+    INSTANCE;
 
-  public QgateProjectFinder(QualityGateDao qualitygateDao, ProjectQgateAssociationDao associationDao) {
-    this.qualitygateDao = qualitygateDao;
-    this.associationDao = associationDao;
+    @Override
+    public Long apply(@Nonnull ProjectQgateAssociationDto input) {
+      return input.getId();
+    }
   }
 
-  public Association find(ProjectQgateAssociationQuery query) {
-    Long gateId = validateId(query.gateId());
-    int pageSize = query.pageSize();
-    int pageIndex = query.pageIndex();
-
-    int offset = (pageIndex - 1) * pageSize;
-    // Add one to page size in order to be able to know if there's more results or not
-    int limit = pageSize + 1;
-    List<ProjectQgateAssociationDto> dtos = associationDao.selectProjects(query, gateId, offset, limit);
-    boolean hasMoreResults = false;
-    if (dtos.size() == limit) {
-      hasMoreResults = true;
-      // Removed last entry as it's only need to know if there more results or not
-      dtos.remove(dtos.size() - 1);
+  private static class MatchProjectId implements Predicate<ProjectQgateAssociationDto> {
+    private final Collection<Long> projectIds;
+
+    private MatchProjectId(Collection<Long> projectIds) {
+      this.projectIds = projectIds;
     }
-    return new Association(toProjectAssociations(dtos), hasMoreResults);
-  }
 
-  private Long validateId(String gateId) {
-    QualityGateDto qualityGateDto = qualitygateDao.selectById(Long.valueOf(gateId));
-    if (qualityGateDto == null) {
-      throw new NotFoundException("Quality gate '" + gateId + "' does not exists.");
+    @Override
+    public boolean apply(@Nonnull ProjectQgateAssociationDto input) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)